src/test/java/org/javarosa/core/model/DynamicSelectUpdateTest.java

```diff
@@ -0,0 +1,398 @@
+            head(
+                title("Repeat trigger"),
+                model(
+                    mainInstance(
+                        t("data id='repeat-trigger'",
+                            t("repeat",
+                                t("question")),
+                            t("select"))),
+
+                    instance("choices",
+                        item("1", "A"),
+                        item("2", "AA"),
+                        item("3", "B"),
+                        item("4", "BB")))),
+                body(
+                    repeat("/data/repeat",
+                        input("/data/repeat/question")),
+                    select1Dynamic("/data/select", "instance('choices')/root/item[value>count(/data/repeat)]")
+                )
+            ));
+
+        scenario.answer("/data/repeat[0]/question", "a");
+        assertThat(scenario.choicesOf("/data/select").size(), is(3));
+
+        scenario.answer("/data/repeat[1]/question", "b");
+        List<SelectChoice> choices = scenario.choicesOf("/data/select");
+        assertThat(choices.size(), is(2));
+        // Because of the repeat trigger in the count expression, choices should be recomputed every time they're requested
+        assertThat(scenario.choicesOf("/data/select"), not(sameInstance(choices)));
+    }
+
+    //region Caching for selects in repeat
+    // When a dynamic select is in a repeat, the itemsets for all repeat instances are represented by the same ItemsetBinding.
+    @Test
+    public void selectInRepeat_withRefToRepeatChildInPredicate_evaluatesChoiceListForEachRepeatInstance() throws Exception {
+        Scenario scenario = Scenario.init("Select in repeat", html(
+            head(
+                title("Select in repeat"),
+                model(
+                    mainInstance(
+                        t("data id='repeat-select'",
+                            t("repeat",
+                                t("select"),
+                                t("filter")))),
+
+                    instance("choices",
+                        item("a", "A"),
+                        item("aa", "AA"),
+                        item("b", "B"),
+                        item("bb", "BB")))),
+            body(
+                repeat("/data/repeat",
+                    select1Dynamic("/data/repeat/select", "instance('choices')/root/item[starts-with(value,current()/../filter)]"),
+                    input("filter"))
+            )));
+
+        scenario.answer("/data/repeat[0]/filter", "a");
+        scenario.answer("/data/repeat[1]/filter", "a");
+        List<SelectChoice> repeat0Choices = scenario.choicesOf("/data/repeat[0]/select");
+        List<SelectChoice> repeat1Choices = scenario.choicesOf("/data/repeat[1]/select");
+
+        // The trigger keys are /data/repeat[0]/filter and /data/repeat[1]/filter which means no caching between them
+        assertThat(repeat0Choices, not(sameInstance(repeat1Choices)));
```

Comment:
Yes. New logging confirms this as the reason the cache was not used:

```
14:26:45.534 [Test worker] DEBUG org.javarosa.core.model.ItemsetBinding - Not using cached value because:
14:26:45.534 [Test worker] DEBUG org.javarosa.core.model.ItemsetBinding -   triggers differ: {/data/repeat[2]/filter=StringData{s='a'}}, {/data/repeat[1]/filter=StringData{s='a'}}
```